lms/djangoapps/courseware/model_data.py

```diff
@@ -34,12 +35,473 @@ class InvalidWriteError(Exception):
+
+        for aside_type in aside_types:
+            usage_ids.add(AsideUsageKeyV1(descriptor.scope_ids.usage_id, aside_type))
+
+    return usage_ids
+
+
+def _all_block_types(descriptors, aside_types):
+    """
+    Return a set of all block_types that are cached by this FieldDataCache.
+    """
+    block_types = set()
+    for descriptor in descriptors:
+        block_types.add(BlockTypeKeyV1(descriptor.entry_point, descriptor.scope_ids.block_type))
+
+    for aside_type in aside_types:
+        block_types.add(BlockTypeKeyV1(XBlockAside.entry_point, aside_type))
+
+    return block_types
+
+
+class DjangoKeyValueStore(KeyValueStore):
+    """
+    This KeyValueStore will read and write data in the following scopes to django models
+        Scope.user_state_summary
+        Scope.user_state
+        Scope.preferences
+        Scope.user_info
+
+    Access to any other scopes will raise an InvalidScopeError
+
+    Data for Scope.user_state is stored as StudentModule objects via the django orm.
+
+    Data for the other scopes is stored in individual objects that are named for the
+    scope involved and have the field name as a key
+
+    If the key isn't found in the expected table during a read or a delete, then a KeyError will be raised
+    """
+
+    _allowed_scopes = (
+        Scope.user_state_summary,
+        Scope.user_state,
+        Scope.preferences,
+        Scope.user_info,
+    )
+
+    def __init__(self, field_data_cache):
+        self._field_data_cache = field_data_cache
+
+    def get(self, key):
+        if key.scope not in self._allowed_scopes:
+            raise InvalidScopeError(key)
+
+        return self._field_data_cache.get(key)
+
+    def set(self, key, value):
+        """
+        Set a single value in the KeyValueStore
+        """
+        self.set_many({key: value})
+
+    def set_many(self, kv_dict):
+        """
+        Provide a bulk save mechanism.
+
+        `kv_dict`: A dictionary of dirty fields that maps
+          xblock.KvsFieldData._key : value
+
+        """
+        for key in kv_dict:
+            # Check key for validity
+            if key.scope not in self._allowed_scopes:
+                raise InvalidScopeError(key)
+
+        self._field_data_cache.set_many(kv_dict)
+
+    def delete(self, key):
+        if key.scope not in self._allowed_scopes:
+            raise InvalidScopeError(key)
+
+        self._field_data_cache.delete(key)
+
+    def has(self, key):
+        if key.scope not in self._allowed_scopes:
+            raise InvalidScopeError(key)
+
+        return self._field_data_cache.has(key)
+
+
+new_contract("DjangoKeyValueStore", DjangoKeyValueStore)
+new_contract("DjangoKeyValueStore_Key", DjangoKeyValueStore.Key)
+
+
+class DjangoOrmFieldCache(object):
+    def __init__(self):
+        self._cache = {}
+
+    def cache_fields(self, fields, descriptors, aside_types):
+        for field_object in self._read_objects(fields, descriptors, aside_types):
```

Comment:
Ah - I guess this is the base class. Could we establish a coding standard whereas any class that's intended as a base class only must have `ABCMeta` as its metaclass? For these reasons:
1. Readability.
2. Inability to instantiate it directly.